From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Banana Pi Maintainers <support@banana-pi.org>
Date: Fri, 22 May 2026 10:00:00 +0800
Subject: [PATCH] sunxi: add Banana Pi P2 Zero

Banana Pi P2 Zero is derived from BPI-M2-Zero and adds eMMC and
100M Ethernet. Add a dedicated U-Boot target so Armbian images can boot
with the matching default DTB and access the extra MMC slot.
---
 arch/arm/dts/Makefile                         |  1 +
 .../dts/sun8i-h2-plus-bananapi-p2-zero.dts    | 39 ++++++++++++++++++++
 configs/bananapi_p2_zero_defconfig            | 13 +++++++
 3 files changed, 53 insertions(+)
 create mode 100644 arch/arm/dts/sun8i-h2-plus-bananapi-p2-zero.dts
 create mode 100644 configs/bananapi_p2_zero_defconfig

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 9d28a485bec6..9d28a485bec6 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -775,5 +775,6 @@ dtb-$(CONFIG_MACH_SUN8I_H3) += \
 dtb-$(CONFIG_MACH_SUN8I_H3) += \
 	sun8i-h2-plus-bananapi-m2-zero.dtb \
+	sun8i-h2-plus-bananapi-p2-zero.dtb \
 	sun8i-h2-plus-libretech-all-h3-cc.dtb \
 	sun8i-h2-plus-orangepi-r1.dtb \
 	sun8i-h2-plus-orangepi-zero.dtb \
diff --git a/arch/arm/dts/sun8i-h2-plus-bananapi-p2-zero.dts b/arch/arm/dts/sun8i-h2-plus-bananapi-p2-zero.dts
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/arch/arm/dts/sun8i-h2-plus-bananapi-p2-zero.dts
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Banana Pi BPI-P2-Zero is based on BPI-M2-Zero and adds eMMC and
+ * 100M Ethernet.
+ */
+
+#include "sun8i-h2-plus-bananapi-m2-zero.dts"
+
+/ {
+	model = "Banana Pi BPI-P2-Zero";
+	compatible = "sinovoip,bpi-p2-zero", "sinovoip,bpi-m2-zero",
+		     "allwinner,sun8i-h2-plus";
+
+	aliases {
+		ethernet0 = &emac;
+	};
+};
+
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	allwinner,leds-active-low;
+	status = "okay";
+};
+
+&mmc2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc2_8bit_pins>;
+	vmmc-supply = <&reg_vcc3v3>;
+	bus-width = <8>;
+	non-removable;
+	cap-mmc-hw-reset;
+	status = "okay";
+};
+
+&mmc2_8bit_pins {
+	drive-strength = <40>;
+	bias-pull-up;
+};
diff --git a/configs/bananapi_p2_zero_defconfig b/configs/bananapi_p2_zero_defconfig
new file mode 100644
index 000000000000..222222222222
--- /dev/null
+++ b/configs/bananapi_p2_zero_defconfig
@@ -0,0 +1,13 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_DEFAULT_DEVICE_TREE="sun8i-h2-plus-bananapi-p2-zero"
+CONFIG_SPL=y
+CONFIG_MACH_SUN8I_H3=y
+CONFIG_DRAM_CLK=408
+# CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
+CONFIG_SUN8I_EMAC=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_MUSB_GADGET=y
+CONFIG_VIDEO_DE2=n
-- 
2.43.0
